src/rust/ide/view/src/project.rs

```diff
@@ -236,6 +230,48 @@ impl Model {
         node_id
     }
 
+    /// Ads a new node below `parent` and returns its ID. If there is not enough space right below
+    /// `parent` then the new node is moved to the right to first gap that is large enough.
+    pub fn add_node_below(&self, parent:NodeId) -> NodeId {
+        let parent_pos = self.graph_editor.model.get_node_position(parent).unwrap_or_default();
+        let styles     = StyleWatch::new(&self.app.display.scene().style_sheet);
+        use ensogl_theme::project as theme;
+        let x_gap    = styles.get_number(theme::default_x_gap_between_nodes);
+        let y_gap    = styles.get_number(theme::default_y_gap_between_nodes);
+        let y_offset = y_gap + node::HEIGHT;
+        let mut x    = parent_pos.x;
+        let y        = parent_pos.y - y_offset;
+        // Push x to the right until we find a position where we have enough space for the new
+        // node, including a margin of size `x_gap`/`y_gap` on all sides.
+        {
+            let nodes             = self.graph_editor.model.nodes.all.raw.borrow();
+            // We subtract `f32::EPSILON` to be sure that we do not include the selected node.
+            let maybe_overlapping = nodes.values().filter(|node|
+                (node.position().y - y).abs() < y_offset - f32::EPSILON);
+            let maybe_overlapping = maybe_overlapping.sorted_by_key(|n|
+                OrderedFloat(n.position().x));
+            // This is how much horizontal space we are looking for.
+            let assumed_width = styles.get_number(theme::assumed_node_width);
+            for node in maybe_overlapping {
+                let node_left  = node.position().x - x_gap;
+                let node_right = node.position().x + node.view.model.width() + x_gap;
+                if x + assumed_width > node_left {
+                    x = x.max(node_right);
+                } else {
+                    // Since `maybe_overlapping` is sorted, we know that the if-condition will
+                    // be false for all following `node`s as well. Therefore, we can skip the
+                    // remaining iterations.
+                    break;
+                }
+            }
+        }
+        let pos = Vector2(x,y);
+        self.graph_editor.add_node.emit(());
+        let node_id = self.graph_editor.frp.output.node_added.value();
+        self.graph_editor.set_node_position((node_id,pos));
+        node_id
+    }
+
```

Comment:
Why this is not a function of graph editor? Graph editor is responsible for creating new nodes, it should be its responsibility to find the place as well IMO

Reply:
I just left it here because the position was also computed here before (It was much simpler at that time). I suppose it could easily be defined as a method on `GraphEditor`. I will try that out.

Reply:
I just noticed that `default_y_gap_between_nodes` is defined in the `project` part of the theme as well. Should that also be moved to the graph-editor part then?

Reply:
yes, it seems so